### PR TITLE
fix(runtime): transcode .oga to .ogg before Whisper transcription

### PR DESCRIPTION
**Pull request by f-liva**
## Problem

Telegram voice notes arrive as a `.oga` payload with MIME `audio/oga` — Telegram's idiosyncratic spelling for Ogg/Opus. Both Whisper variants LibreFang auto-detects (`groq/whisper-large-v3-turbo` and `openai/whisper-1`) fail their format probe on `.oga`: the request is accepted, but transcription returns garbage or surfaces as "Audio transcription failed: Unsupported format". Observed repeatedly in production (`WARN librefang_api::routes::agents: Audio transcription failed: ...` against `audio.ogg` → file_id ... — Telegram voice), despite the file being a perfectly valid Ogg/Opus stream.

The previous inline MIME → extension map in `MediaEngine::transcribe_audio` didn't know about `audio/oga` and silently bucketed it as `"wav"`, which made the situation worse.

## Fix

Re-encode `.oga` → `.ogg` via ffmpeg right before the Whisper upload. `libopus` keeps the Opus payload bit-exact (no lossy re-encode — just re-packetisation), and the outbound MIME / filename become the plain `audio/ogg` / `audio.ogg` that Whisper recognises.

### Implementation highlights

- **New `transcode_oga_to_ogg_opus(bytes)`** in `crates/librefang-runtime/src/media_understanding.rs`:
  - Pipes input on `stdin` and reads output from `stdout` — no scratch files. Telegram voice notes are typically <1 MB, comfortably under the pipe buffer.
  - Two tokio tasks drain `stdout` / `stderr` concurrently so a full buffer on either side doesn't deadlock the transcode.
  - `wait()` under a 30 s wall-clock timeout. On timeout we explicitly `start_kill()` + `wait()` to reap the child — avoids zombie ffmpeg processes under sustained load.
- **New `mime_to_ext` helper** centralises the MIME → Whisper-accepted extension mapping (was inline in `transcribe_audio`). Unknown MIMEs now fall back to the source file's own extension rather than hard-coding `"wav"`.
- Trigger: `ext == "oga"` OR MIME matches `audio/oga` (case-insensitive).
- Transcode failure is a hard error (`ffmpeg .oga transcode failed: …`). Silently passing the raw `.oga` through with a renamed extension is the exact case the bug originally produces, so there's no honest fallback — surfacing an "install ffmpeg" style error is the correct UX.

## Testing

All tests skip cleanly when `ffmpeg` isn't on `PATH` (CI images and the production container both have it; dev machines may not):

- `mime_to_ext_maps_known_types` — every entry that used to live as inline match arms.
- `mime_to_ext_is_case_insensitive` — RFC says MIME is case-insensitive.
- `mime_to_ext_returns_none_for_unmapped` — pins that `audio/oga` stays unmapped so the caller takes the transcode branch.
- `transcode_oga_smoke` — synthesises 0.5 s of Opus silence via ffmpeg, round-trips through the transcoder, asserts the output starts with the `OggS` magic marker.
- `transcode_empty_input_errors` — exercises the ffmpeg failure path.

## Attribution

- [x] Original work

### CODE REVIEW

File: crates/librefang-runtime/src/media_understanding.rs [hunk 1]
```diff
@@ -266,6 +292,110 @@ fn detect_vision_provider() -> Option<&'static str> {
...
+        .args([
+            "-hide_banner",
+            "-loglevel",
+            "error",
+            "-f",
+            "ogg",
+            "-i",
+            "pipe:0",
+            "-vn",
+            "-c:a",
```
Comment (houko) [inline crates/librefang-runtime/src/media_understanding.rs:329]: **Code/body mismatch + technical concern.** The PR body says 'libopus keeps the Opus payload bit-exact (no lossy re-encode — just re-packetisation)' but `-c:a libopus` does the opposite: it decodes the Opus stream to PCM and re-encodes to Opus. That's lossy (Opus is a lossy codec, re-encoding compounds quantization), slower, and requires `libopus` to be compiled into the local `ffmpeg` — which is not guaranteed on minimal Docker images. If you want bit-exact re-packetisation, use `-c:a copy`:

```diff
             "-f",
             "ogg",
             "-i",
             "pipe:0",
             "-vn",
             "-c:a",
-            "libopus",
+            "copy",
             "-f",
             "ogg",
             "pipe:1",
```

Benefits, in order:

1. **Actually bit-exact.** The Opus payload is passed through unchanged — ffmpeg just re-muxes into a new Ogg container. This matches what the PR body claims.
2. **Strictly faster.** No decode, no re-encode. For a <1 MB Telegram voice note the difference is milliseconds, but it also means the 30 s timeout budget is effectively bottomless.
3. **Works with any ffmpeg build.** Minimal Alpine-based Docker images (including some of the base images listed in `Dockerfile`s I've seen in the ecosystem) ship a reduced `ffmpeg` without `libopus`. With `-c:a copy` you only need the Ogg demuxer + muxer, both of which are in every ffmpeg build.
4. **No quality loss on repeat.** If the same file is ever re-processed (retry path, cache invalidation, session repair re-reads the attachment), successive re-encodes accumulate quality loss with `libopus`. With `copy` you get the same bytes out every time.

One subtle thing to verify if you switch: the smoke test at L502 also uses `-c:a libopus` — but that's the *test fixture* generating synthetic audio from `anullsrc`, which is PCM and genuinely needs encoding. So only the transcoder call at L329 needs the flag change; the test can stay as-is.

**Review by houko — DISMISSED** [review]
Clean, targeted PR — single file, real observed production bug (Telegram `audio/oga` → garbage transcription), comprehensive test coverage including the pipe-deadlock edge case, and everything skips cleanly when `ffmpeg` isn't on PATH. Local `cargo check -p librefang-runtime` is warning-free. I'd merge this as-is, **except for one technical point where the code doesn't match what the PR body promises** — details inline.

## The issue

The PR body says:

> `libopus` keeps the Opus payload bit-exact (no lossy re-encode — just re-packetisation)

but the ffmpeg args at `media_understanding.rs:329` specify `-c:a libopus`. That's a **full decode+re-encode**, not stream copy. For bit-exact re-packetisation you want `-c:a copy`. See the inline comment for the impact — it's a one-word fix and it's strictly better on every axis (quality, speed, build compatibility with minimal ffmpeg images, memory).

If you'd rather land this as-is and fix the codec flag in a followup, that's defensible — the current code *works*, it's just doing more work than it claims to, and it creates a dependency on libopus being compiled into whatever ffmpeg the deployment ships. Given this is the kind of fix that's going to sit untouched for months once merged, I'd prefer to get it right now.

## Everything else

Positive observations, no action needed:

- **`mime_to_ext` extraction is the right refactor.** Moving the mapping out of `transcribe_audio` and adding explicit case-insensitive handling + `None` for unknown types is cleaner than the old inline match with a silent `"wav"` fallback. The tests cover the full mapping including the deliberate `audio/oga → None` invariant.
- **Pipe handling is correct.** Feeding stdin from a `tokio::spawn` task + draining stdout/stderr in separate tasks before `wait()` is the standard pattern for avoiding pipe-buffer deadlocks. I checked: the tasks read until EOF from the (possibly-killed) child's pipes, and they finish on their own after timeout — no leaks.
- **Timeout + reap handling is correct.** `start_kill()` + `wait().await` on timeout avoids zombie ffmpeg processes, which is important under sustained Telegram voice-note load.
- **Hard-error on transcode failure is the right UX.** The silent passthrough to raw `.oga` was the bug; making operators see 'install ffmpeg' is strictly better than letting Whisper silently return garbage.
- **Test coverage pins the right invariants:** `audio/oga` is explicitly asserted to stay unmapped so the transcode branch fires; the smoke test generates real Opus via `anullsrc` and verifies the `OggS` magic on output.

## Minor (non-blocking)

- **`failed to spawn ffmpeg: {e}`** (L314) — when ffmpeg isn't installed, `e` is `No such file or directory (os error 2)`, which isn't actionable. Consider: `"ffmpeg not found — install it (brew install ffmpeg / apt install ffmpeg) to process Telegram voice notes"`.
- **Stdin writer errors are silently swallowed** (L336-340). That's actually correct here — if stdin writing fails, ffmpeg exits with a non-zero status and stderr has the real reason. A one-line comment explaining why the `let _ =` is intentional would help future readers not flag it as a lint.
- **`transcode_empty_input_errors` accepts either branch** (`contains("ffmpeg exited") || contains("empty output")`). It's lenient but that's fine — whichever branch ffmpeg takes is implementation-defined. Could split into two tests (one with `[]`, one with random non-Ogg bytes) if you want tighter coverage.

Happy to approve once the codec flag is swapped, or to approve as-is if you prefer to land and fix in a followup. Your call.

**Comment by f-liva** [comment]
Follow-up PR #2459 opened to address the three non-blocking observations from your review: actionable ffmpeg-missing error message, comment on the intentional `let _ =` on the stdin writer, and split of `transcode_empty_input_errors` into two tests (empty input vs. non-Ogg junk) that each pin a single branch. Thanks again for the thorough review.
